File: sdk/elixir/examples/03-containers/sykli.exs
# Example 03: Container Execution
#
# This example demonstrates:
# - `container` for isolated execution
# - `dir` and `cache` resources
# - `mount` and `mount_cache` for volumes
# - `workdir` for working directory
# - `env` for environment variables
#
# Run with: elixir sykli.exs --emit | sykli run -

Mix.install([{:sykli, "~> 0.2"}])

use Sykli

pipeline do
  # === RESOURCES ===

  # Source directory - mounted into containers
  src = dir(".")

  # Named caches - persist across runs
  deps_cache = cache("mix-deps")
  build_cache = cache("mix-build")

  # === TASKS ===

  # Lint in container with source mounted
  task "lint" do
    container "elixir:1.16"
    mount src, "/app"
    mount_cache deps_cache, "/app/deps"
    workdir "/app"
    run "mix credo --strict"
  end

  # Test with multiple caches
  task "test" do
    container "elixir:1.16"
    mount src, "/app"
    mount_cache deps_cache, "/app/deps"
    mount_cache build_cache, "/app/_build"
    workdir "/app"
    env "MIX_ENV", "test"
    run "mix test"
  end

  # Build with environment variable
  task "build" do
    container "elixir:1.16"
    mount src, "/app"
    mount_cache deps_cache, "/app/deps"
    mount_cache build_cache, "/app/_build"
    workdir "/app"
    env "MIX_ENV", "prod"
    run "mix release"
    output "release", "_build/prod/rel/myapp"
    after_ ["lint", "test"]
  end

  # === CONVENIENCE METHODS ===

  # mount_cwd() is shorthand for mounting . to /work
  task "check-format" do
    container "elixir:1.16"
    mount_cwd()  # Mounts current dir to /work, sets workdir
    run "mix format --check-formatted"
  end
end

# Key concepts:
# - dir(".")      - Represents a host directory
# - cache("name") - Named persistent cache volume
# - mount(d, p)   - Mount directory at container path
# - mount_cache() - Mount cache for faster builds
# - workdir()     - Set container working directory

File: sdk/go/examples/03-containers/sykli.go
//go:build ignore

// Example 03: Container Execution
//
// This example demonstrates:
// - Container() for isolated execution
// - Dir() and Cache() resources
// - Mount() and MountCache() for volumes
// - Workdir() for working directory
// - Env() for environment variables
//
// Run with: sykli run

package main

import sykli "github.com/yairfalse/sykli/sdk/go"

func main() {
	s := sykli.New()

	// === RESOURCES ===

	// Source directory - mounted into containers
	src := s.Dir(".")

	// Named caches - persist across runs
	goModCache := s.Cache("go-mod")       // Go module cache
	goBuildCache := s.Cache("go-build")   // Go build cache

	// === TASKS ===

	// Lint in container with source mounted
	s.Task("lint").
		Container("golang:1.21").
		Mount(src, "/src").
		MountCache(goModCache, "/go/pkg/mod").
		Workdir("/src").
		Run("go vet ./...")

	// Test with multiple caches
	s.Task("test").
		Container("golang:1.21").
		Mount(src, "/src").
		MountCache(goModCache, "/go/pkg/mod").
		MountCache(goBuildCache, "/root/.cache/go-build").
		Workdir("/src").
		Run("go test -v ./...")

	// Build with environment variable
	s.Task("build").
		Container("golang:1.21").
		Mount(src, "/src").
		MountCache(goModCache, "/go/pkg/mod").
		MountCache(goBuildCache, "/root/.cache/go-build").
		Workdir("/src").
		Env("CGO_ENABLED", "0").
		Env("GOOS", "linux").
		Run("go build -o ./app").
		Output("binary", "./app").
		After("lint", "test")

	// === CONVENIENCE METHODS ===

	// MountCwd() is shorthand for mounting . to /work
	s.Task("check-format").
		Container("golang:1.21").
		MountCwd().  // Mounts current dir to /work, sets workdir
		Run("gofmt -l .")

	s.Emit()
}

// Key concepts:
// - Dir(".")      - Represents a host directory
// - Cache("name") - Named persistent cache volume
// - Mount(d, p)   - Mount directory at container path
// - MountCache()  - Mount cache for faster builds
// - Workdir()     - Set container working directory
